flow: gopacket OpenFlow decoder

lebauce opened the pull request.

lebauce commented:
run skydive-functional-tests-backend-elasticsearch

pierrecregut commented:
I will not manage to make a real review this week. Here are a few thoughts:
* It looks very promissing as it seems that floodlight has an up-to-date openflow library. But what is the level of confidence on loxigen maintenance ? There are not a lot of commits on both floodlight and loxigen projects and not from the initial developers.
* The real core of the proposal is not in skydive but in your library and in fact in the generator used for your library goloxi (golang branch of your loxigen fork). Do you intend to push your changes to loxigen ?
Why don't you push all those projects under skydive-project umbrella ?
* It seems to generate 6 silos: one for each version of OpenFlow. Is there not a risk that it will be hard to use ? Usually you announce you support a lot of versions (you only blacklist the version known to have problems) and the switch replies with whatever it supports. Go and Java do not have the same notions of type hierarchies, it may be easier in Go in fact.

Otherwise, I tried it on my regular demo environment (standard devstack single node with ml2/ovs). It quickly crashes with:

```
panic: runtime error: slice bounds out of range

goroutine 91 [running]:
github.com/skydive-project/skydive/vendor/github.com/skydive-project/goloxi.(*Decoder).SliceDecoder(...)
        /home/piac6784/golang/src/github.com/skydive-project/skydive/vendor/github.com/skydive-project/goloxi/decoder.go:90
github.com/skydive-project/skydive/vendor/github.com/skydive-project/goloxi/of10.(*Header).Decode(0xc42007de50, 0xc4200f1170, 0x0, 0x0)
        /home/piac6784/golang/src/github.com/skydive-project/skydive/vendor/github.com/skydive-project/goloxi/of10/message.go:87 +0x325
github.com/skydive-project/skydive/openflow.(*Client).readLoop(0xc4205a0e10)
        /home/piac6784/golang/src/github.com/skydive-project/skydive/openflow/client.go:171 +0x2ed
created by github.com/skydive-project/skydive/openflow.(*Client).Start
        /home/piac6784/golang/src/github.com/skydive-project/skydive/openflow/client.go:266 +0x152
```

May be we should quickly distinguish issues in loxigen/goloxi and issues in its integration in skydive.
The first question is how to validate goloxi ? Do they have a functional test suite ? I see a test_data folder. How to use it with goloxi ?

lebauce commented:
> * It looks very promissing as it seems that floodlight has an up-to-date openflow library. But what is the level of confidence on loxigen maintenance ? There are not a lot of commits on both floodlight and loxigen projects and not from the initial developers.

Loxigen is currently maintained (at a slow pace), but as of now, it's up to date. Every OpenFlow specifications use the same mechanisms so adding support for a new version is not an easy task but it's perfectly feasible in a reasonable time. And even if Loxigen is discontinued we still have OpenFlow 1.5. On my Fedora box, OVS does not enable OpenFlow 1.5 support by default.

> * The real core of the proposal is not in skydive but in your library and in fact in the generator used for your library goloxi (golang branch of your loxigen fork). Do you intend to push your changes to loxigen ?
As you said, almost everything is the Golang loxigen backend. I will push my changes to loxigen for sure. What makes it a bit hard is that there are 3 types of changes to loxigen:
- Fixes to the loxigen definitions (wrong types, missing enums, ...)
- Support for Nicira extension (NXM and NxActions). (This was done by parsing the Openvswitch header files and manually checking that this was done properly).
- Some changes I had to make to the loxigen core to be able to describe some Nicira NXM or actions. Some of them are really exotic and did not fit into the loxigen model.

While the loxigen guys may accept the 2 first, the last one is more problematic as it imports all the backends only for Nicira extensions. The changes are fairly small so it's ok to carry these in our fork.

>   Why don't you push all those projects under skydive-project umbrella ?
I wasn't sure about the `goloxi` option when I started it. So I put it under my account. But it will definitely be moved under the skydive-project umbrella when we decide to merge it.

> * It seems to generate 6 silos: one for each version of OpenFlow. Is there not a risk that it will be hard to use ? Usually you announce you support a lot of versions (you only blacklist the version known to have problems) and the switch replies with whatever it supports. Go and Java do not have the same notions of type hierarchies, it may be easier in Go in fact.
It's true. But this is mainly because of OpenFlow itself : every OpenFlow versions comes with its set of objects and concepts, often reused from the previous one but with some changes. A same message type can have different id depending on the OpenFlow version (there are only 8 immutable message types). That's why all the Golang libraries I found supported only one OpenFlow version.
That being said, I tried to make this a bit easier to use by defining Go interfaces for every object and defining getters and setters for every field. So you can access an OpenFlow object the same way between OpenFlow versions.

> 
> Otherwise, I tried it on my regular demo environment (standard devstack single node with ml2/ovs). It quickly crashes with:
> 

I removed the wip tag way to early. I was talking to the OVS using OpenFlow 1.4 and only OpenFlow 1.3 is enabled by default on devstack. So I'm currently modifying my PR so that we negotiate the OpenFlow version to use and use the newest one supported by the switch.

> ```
> panic: runtime error: slice bounds out of range
> 
> goroutine 91 [running]:
> github.com/skydive-project/skydive/vendor/github.com/skydive-project/goloxi.(*Decoder).SliceDecoder(...)
>         /home/piac6784/golang/src/github.com/skydive-project/skydive/vendor/github.com/skydive-project/goloxi/decoder.go:90
> github.com/skydive-project/skydive/vendor/github.com/skydive-project/goloxi/of10.(*Header).Decode(0xc42007de50, 0xc4200f1170, 0x0, 0x0)
>         /home/piac6784/golang/src/github.com/skydive-project/skydive/vendor/github.com/skydive-project/goloxi/of10/message.go:87 +0x325
> github.com/skydive-project/skydive/openflow.(*Client).readLoop(0xc4205a0e10)
>         /home/piac6784/golang/src/github.com/skydive-project/skydive/openflow/client.go:171 +0x2ed
> created by github.com/skydive-project/skydive/openflow.(*Client).Start
>         /home/piac6784/golang/src/github.com/skydive-project/skydive/openflow/client.go:266 +0x152
> ```

> May be we should quickly distinguish issues in loxigen/goloxi and issues in its integration in skydive.
You're right. The MarshalJSON part of `goloxi` is already pretty much only to have nice JSON objects for Skydive. It's completely generic though.

> The first question is how to validate goloxi ? Do they have a functional test suite ? I see a test_data folder. How to use it with goloxi ?
There is indeed a test suite for goloxi and it needs to be adapted for every backend. It's definitely on my TODO list (it's also probably a requirement because pushing our loxigen changes), but I did not have time for it yet.

lebauce commented:
@pierrecregut I updated my PR to use the OpenFlow version specified by the OVS switch. I also fixed a bug related to padding of simple actions. Could you please give it a try ? Thanks

lebauce commented:
run skydive-functional-tests-backend-orientdb

lebauce commented:
run skydive-functional-tests-backend-elasticsearch

lebauce commented:
run skydive-functional-tests-backend-orientdb

lebauce commented:
skip skydive-cdd-overview-tests

lebauce commented:
run skydive-cdd-overview-tests

lebauce commented:
run skydive-functional-tests-backend-elasticsearch

lebauce commented:
run skydive-functional-tests-backend-orientdb

lebauce commented:
run skydive-functional-tests-backend-elasticsearch

lebauce commented:
run skydive-scale-tests

lebauce commented:
run skydive-functional-tests-backend-orientdb

safchain commented:
@pierrecregut We would like to merge this one, but before we wanted to be sure that is OK for you ?

lebauce commented on topology/probes/ovsdb/ovs_of.go:
```diff
@@ -46,814 +30,57 @@ import (
+	address, ok := o.Translation[bridge]
+	if !ok {
+		logging.GetLogger().Warningf("Could not find translation address for %s in %v", bridge, o.Translation)
+		address = fmt.Sprintf("unix:/var/run/openvswitch/%s.mgmt", bridge)
```
Indeed

lebauce commented:
run skydive-cdd-overview-tests

lebauce commented:
run skydive-functional-tests-backend-orientdb

lebauce commented:
run skydive-functional-tests-backend-orientdb

lebauce commented:
run skydive-functional-tests-backend-orientdb

lebauce commented:
run skydive-compile-tests

lebauce commented:
run skydive-functional-hw-tests

lebauce commented:
run skydive-functional-tests-backend-orientdb

lebauce commented:
run skydive-functional-tests-backend-orientdb

safchain commented:
let's merge it and improve it (if needed :)
